jb/itch5/compute_book.hpp

```diff
@@ -19,91 +19,96 @@
-     */
-    time_point recvts;
+  stock_t stock;
 
-    /**
-     * The security updated by this order.  This is redundant for
-     * order updates and deletes, and ITCH-5.0 omits the field, but
-     * we find it easier
-     */
-    stock_t stock;
+  /// What side of the book is being updated.
+  buy_sell_indicator_t buy_sell_indicator;
 
-    /// What side of the book is being updated.
-    buy_sell_indicator_t buy_sell_indicator;
+  /// What price level is being updated.
+  price4_t px;
 
-    /// What price level is being updated.
-    price4_t px;
+  /// How many shares are being added (if positive) or removed (if
+  /// negative) from the book.
+  int qty;
 
-    /// How many shares are being added (if positive) or removed (if
-    /// negative) from the book.
-    int qty;
+  /// If true, this was a cancel replace and and old order was
+  /// modified too...
+  bool cxlreplx;
 
-    /// If true, this was a cancel replace and and old order was
-    /// modified too...
-    bool cxlreplx;
+  /// Old price for the order
+  price4_t oldpx;
 
-    /// Old price for the order
-    price4_t oldpx;
+  /// How many shares were removed in the old order
+  int oldqty;
+};
 
-    /// How many shares were removed in the old order
-    int oldqty;
-  };
+/**
+  * A convenient container for per-order data.
+  *
+  * Most market data feeds resend the security identifier and side
+  * with each order update, but ITCH-5.0 does not.  One needs to
+  * lookup the symbol, side, original price,  information based on the order
+  * id.  This literal type
+  * is used to keep that information around.
+  */
+struct order_data {
+  /// The symbol for this particular order
+  stock_t stock;
+
+  /// Whether the order is a BUY or SELL
+  buy_sell_indicator_t buy_sell_indicator;
+
+  /// The price of the order
+  price4_t px;
+
+  /// The remaining quantity in the order
+  int qty;
+};
 
+/**
+ * Compute the book and call a user-defined callback on each change.
+ *
+ * Keep a collection of all the order books, indexed by symbol, and
+ * forward the updates to them.  Only process the relevant messages
+ * types in ITCH-5.0 that are necessary to keep the book.
+ */
+template <typename book_type>
+class compute_book {
+public:
+  //@{
   /**
-   * A convenient container for per-order data.
-   *
-   * Most market data feeds resend the security identifier and side
-   * with each order update, but ITCH-5.0 does not.  One needs to
-   * lookup the symbol, side, original price,  information based on the order
-   * id.  This literal type
-   * is used to keep that information around.
+   * @name Type traits
    */
-  struct order_data {
-    /// The symbol for this particular order
-    stock_t stock;
-
-    /// Whether the order is a BUY or SELL
-    buy_sell_indicator_t buy_sell_indicator;
-
-    /// The price of the order
-    price4_t px;
+  /// Define the clock used to measure processing delays
+  using clock_type = jb::itch5::clock_type;
```

Comment:
I had to make these 2 types redundant (clock_type and time_point) because:

1) I took the book_update and order_data struct definition out of the class compute_book, to avoid using them as template class members (besides the code style, they are not dependent on the template parameter). By doing so, I needed the this two types declared before the struct definition.... so I took them our the class two.... however, there are other modules (process_something_something) that I have not touched before using the compute_book::clock_type type...

example:
"
../tools/mold2inside.cpp:114:77:   required from here
../jb/itch5/process_buffer_mlist.hpp:90:15: error: no type named ‘time_point’ in ‘class jb::itch5::compute_book<jb::itch5::map_price>’
"

2) So, I decided to create an alias inside the template class to the type define outside, instead of changing the code where the call actually uses : jb::itch5::compute_book<param>::clock_type (instead of b::itch5::clock_type)

Advice?